### PR TITLE
fix(dot/sync): verify justification before importing blocks

### PR DESCRIPTION
**Pull request by EclesioMeloJunior**
## Changes

- Changed sync to verify block data justification before processing/importing block into the state
- Changed `finalization.VerifyBlockJustification`, now it don't set the finalized block at the end of the function, the function now just verifies and returns the round and `set id`
- The `blockState.SetFinalized` is now called at the end of processing/importing a block together with `blockState.SetJustification` at `finalizeAndSetJustification` method

## Tests

```sh
go test -tags integration github.com/ChainSafe/gossamer/dot/sync
```

## Issues

- Closes https://github.com/ChainSafe/gossamer/issues/3468
- Closes https://github.com/ChainSafe/gossamer/issues/3308

## Primary Reviewer

@timwu20

### CODE REVIEW

File: dot/network/message.go [hunk 1]
```diff
@@ -388,7 +390,7 @@ func NewAscendingBlockRequests(startNumber, targetNumber uint, requestedData byt
 	numRequests := diff / MaxBlocksInResponse
 	// we should check if the diff is in the maxResponseSize bounds
 	// otherwise we should increase the numRequests by one, take this
-	// example, we want to sync from 0 to 259, the diff is 259
+	// example, we want to sync from 1 to 259, the diff is 259
```
Comment (jimjbrettj) [inline dot/network/message.go:393]: would this make the diff 258?

File: lib/runtime/wazero/test_helpers.go [hunk 2]
```diff
@@ -72,6 +72,7 @@ func NewTestInstanceWithTrie(t *testing.T, targetRuntime string, tt *trie.Trie)
 
 // NewInstanceFromFile instantiates a runtime from a .wasm file
 func NewInstanceFromFile(fp string, cfg Config) (*Instance, error) {
+	fmt.Println(fp)
```
Comment (jimjbrettj) [inline lib/runtime/wazero/test_helpers.go:75]: nit/remove print

**Review by jimjbrettj — APPROVED** [review]
just a few comments, but lgtm

File: dot/network/message.go (hunk 1 above)
Comment (dimartiro) [reply dot/network/message.go:393]: since `diff := targetNumber - (startNumber - 1)` I think 259 is right

File: lib/grandpa/message_handler.go [hunk 3]
```diff
@@ -404,60 +404,61 @@ func (h *MessageHandler) verifyPreCommitJustification(msg *CatchUpResponse) erro
 
 // VerifyBlockJustification verifies the finality justification for a block, returns scale encoded justification with
 // any extra bytes removed.
-func (s *Service) VerifyBlockJustification(hash common.Hash, justification []byte) error {
+func (s *Service) VerifyBlockJustification(hash common.Hash, justification []byte) (
+	round uint64, setID uint64, err error) {
```
Comment (kanishkatn) [inline lib/grandpa/message_handler.go:408]: maybe update the comments to reflect the change in return?

File: lib/grandpa/message_handler_integration_test.go [hunk 4]
```diff
@@ -772,8 +772,11 @@ func TestMessageHandler_VerifyBlockJustification_WithEquivocatoryVotes(t *testin
 	just := newJustification(round, testHash, number, precommits)
 	data, err := scale.Marshal(*just)
 	require.NoError(t, err)
-	err = gs.VerifyBlockJustification(testHash, data)
+
+	gotRound, gotSetID, err := gs.VerifyBlockJustification(testHash, data)
 	require.NoError(t, err)
+	require.Equal(t, round, gotRound)
+	require.Equal(t, setID, gotSetID)
```
Comment (kanishkatn) [inline lib/grandpa/message_handler_integration_test.go:779]:
```suggestion
	actualRound, actualSetID, err := gs.VerifyBlockJustification(testHash, data)
	require.NoError(t, err)
	require.Equal(t, round, actualRound)
	require.Equal(t, setID, actualSetID)
```

File: dot/sync/chain_sync.go [hunk 5]
```diff
@@ -782,7 +782,7 @@ func (cs *chainSync) handleReadyBlock(bd *types.BlockData) error {
 	err := cs.processBlockData(*bd)
 	if err != nil {
 		// depending on the error, we might want to save this block for later
-		logger.Errorf("block data processing for block with hash %s failed: %s", bd.Hash, err)
+		logger.Errorf("processing: %s", err)
```
Comment (kanishkatn) [inline dot/sync/chain_sync.go:785]: Is this intentional?

**Review by kanishkatn — APPROVED** [review]

**Pull request merged**